library/std/src/sys/pal/common/exit_guard.rs

```diff
@@ -0,0 +1,98 @@
+cfg_if::cfg_if! {
+    if #[cfg(miri)] {
+        /// Mitigation for <https://github.com/rust-lang/rust/issues/126600>
+        ///
+        /// This mitigation is not necessary when running under Miri, so this function does nothing
+        /// when running under Miri.
+        pub(crate) fn unique_thread_exit() {
+            // Mitigation not required on Miri, where `atexit` is not supported and
+            // hence `exit` is thread-safe.
+        }
```

Comment:
`libc::gettid` is not usable under Miri IIUC, and [`std::thread::current().id().as_u64()` can fail](https://github.com/rust-lang/rust/pull/126606#issuecomment-2176784818) if we are currently exiting (or in  TLS destructor? (etc?)).

Is there a simple way to get a unique thread ID that is both supoorted under Miri, and works at arbitrary times? I suppose a manual atomic counter and thread_local ID that incrments the counter if not yet set would work.

 Unless it is acceptable that calling `exit` at  time when `std::thread::try_current()` fails will abort, but that would then happen even on calls to `exit` that would not otherwise abort.

Reply:
Implementing `gettid` in Miri should not be too hard. But maybe that should not block this PR.

Reply:
@zachs18 Does `pthread_self` work in Miri?

Reply:
Yes, we support `pthread_self` in Miri.

Reply:
Latest push uses `pthread_self` and `Mutex<Option<pthread_t>>` instead of `libc::gettid` and `AtomicI32` for maximum portability[^1]. This also removes the code duplication since there is no "fallback" implementation. Given that this code should not be hot, I think it should be fine to not use atomics here.

One possible issue is that *formally*, comparing `pthread_t`s is supposed to use [`pthread_equal`](https://man7.org/linux/man-pages/man3/pthread_equal.3.html), not `==` comparison, but the `libc` crate currently does not expose `pthread_equal`. I don't know if any supported libcs *actually* make `pthread_equal` do anything  other than `==`, though, so it might just be fine (glibc and musl both appear to implement `pthread_equal` as `==` (glibc) ([musl](https://github.com/bminor/musl/blob/ab31e9d6a0fa7c5c408856c89df2dfb12c344039/include/pthread.h#L92))). Alternately, we could just have a one-off `extern "C" { fn pthread_equal(...) -> ...; }` above the call, at least until/unless `libc` exposes `pthread_equal`.

[^1]:  since `pthread_t` is different on different Linux platforms: `*-unknown-linux-gnu` appears to have it as `c_ulong` (which is `u64` on 64-bit and `u32` on 32-bit), `*-unknown-linux-musl` appears to have it as `*mut c_void`, and `*-linux-android` appears to have it as `c_long`